File: Sources/Domain/ConnectionAttributionCoverage.swift
struct ConnectionMetadataAvailability: Equatable, Sendable {
  let hasHostname: Bool
  let hasApplication: Bool

  static let unavailable = ConnectionMetadataAvailability(
    hasHostname: false,
    hasApplication: false
  )
}

struct ConnectionMetadata: Equatable, Sendable {
  let hostname: String?
  let applicationName: String?

  static let unavailable = ConnectionMetadata(hostname: nil, applicationName: nil)

  var availability: ConnectionMetadataAvailability {
    ConnectionMetadataAvailability(
      hasHostname: hostname != nil,
      hasApplication: applicationName != nil
    )
  }
}

struct ConnectionAttributionCoverage: Equatable, Sendable {
  let proxyConnectionCount: Int
  let hostnameIdentifiedCount: Int
  let applicationIdentifiedCount: Int
  let fullyIdentifiedCount: Int

  static let empty = ConnectionAttributionCoverage(
    proxyConnectionCount: 0,
    hostnameIdentifiedCount: 0,
    applicationIdentifiedCount: 0,
    fullyIdentifiedCount: 0
  )

  var hostnameRate: Double? {
    rate(for: hostnameIdentifiedCount)
  }

  var applicationRate: Double? {
    rate(for: applicationIdentifiedCount)
  }

  var fullyIdentifiedRate: Double? {
    rate(for: fullyIdentifiedCount)
  }

  private func rate(for count: Int) -> Double? {
    guard proxyConnectionCount > 0 else {
      return nil
    }
    return Double(count) / Double(proxyConnectionCount)
  }
}

struct ConnectionAttributionCoverageTracker: Sendable {
  private var availabilityByConnectionID: [String: ConnectionMetadataAvailability] = [:]

  mutating func consume(
    _ connections: [ConnectionTrafficSample]
  ) -> ConnectionAttributionCoverage {
    for connection in connections {
      let previous = availabilityByConnectionID[connection.id] ?? .unavailable
      availabilityByConnectionID[connection.id] = ConnectionMetadataAvailability(
        hasHostname: previous.hasHostname || connection.metadataAvailability.hasHostname,
        hasApplication: previous.hasApplication || connection.metadataAvailability.hasApplication
      )
    }
    return coverage
  }

  mutating func reset() {
    availabilityByConnectionID = [:]
  }

  private var coverage: ConnectionAttributionCoverage {
    let values = Array(availabilityByConnectionID.values)
    return ConnectionAttributionCoverage(
      proxyConnectionCount: values.count,
      hostnameIdentifiedCount: values.filter(\.hasHostname).count,
      applicationIdentifiedCount: values.filter(\.hasApplication).count,
      fullyIdentifiedCount: values.count { $0.hasHostname && $0.hasApplication }
    )
  }
}

File: platform/windows/MihomoMeter.Windows.Core/Domain/ConnectionAttributionCoverage.cs
namespace MihomoMeter.Windows.Core.Domain;

public readonly record struct ConnectionMetadataAvailability(
    bool HasHostname,
    bool HasApplication)
{
    public static ConnectionMetadataAvailability Unavailable => default;
}

public readonly record struct ConnectionAttributionCoverage(
    int ProxyConnectionCount,
    int HostnameIdentifiedCount,
    int ApplicationIdentifiedCount,
    int FullyIdentifiedCount)
{
    public static ConnectionAttributionCoverage Empty => default;

    public double? HostnameRate => RateFor(HostnameIdentifiedCount);

    public double? ApplicationRate => RateFor(ApplicationIdentifiedCount);

    public double? FullyIdentifiedRate => RateFor(FullyIdentifiedCount);

    private double? RateFor(int count)
    {
        return ProxyConnectionCount == 0
            ? null
            : (double)count / ProxyConnectionCount;
    }
}

public sealed class ConnectionAttributionCoverageTracker
{
    private readonly Dictionary<string, ConnectionMetadataAvailability> _availabilityByConnectionId =
        new(StringComparer.Ordinal);

    public ConnectionAttributionCoverage Consume(
        IReadOnlyList<ConnectionTrafficSample> proxyConnections)
    {
        foreach (var connection in proxyConnections)
        {
            _availabilityByConnectionId.TryGetValue(
                connection.Id,
                out var previousAvailability);
            _availabilityByConnectionId[connection.Id] = new ConnectionMetadataAvailability(
                previousAvailability.HasHostname || connection.MetadataAvailability.HasHostname,
                previousAvailability.HasApplication || connection.MetadataAvailability.HasApplication);
        }

        return Coverage;
    }

    public void Reset()
    {
        _availabilityByConnectionId.Clear();
    }

    private ConnectionAttributionCoverage Coverage
    {
        get
        {
            var values = _availabilityByConnectionId.Values;
            return new ConnectionAttributionCoverage(
                values.Count,
                values.Count(value => value.HasHostname),
                values.Count(value => value.HasApplication),
                values.Count(value => value.HasHostname && value.HasApplication));
        }
    }
}
